### PR TITLE
NestedTensor Softmax

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Nested tensor softmax implementation for cuda, based on  mask compute followed by softmax

Test Plan: unit test

Differential Revision: D38288107

### CODE REVIEW

**Review by jbschlosser — CHANGES_REQUESTED** [review]
This PR & the NT transformer code in general do some things that aren't within best practices for PyTorch development, contributing to difficulty passing CI checks.

From what I can tell, these are the cases we're concerned with:

1. Non-NT input, no `attn_mask` specified - should do standard `softmax`
2. Non-NT input, `attn_mask` specified - should call `_masked_softmax`
3. NT input, no `attn_mask` specified - should do `softmax` according to NT nested structure
4. NT input, `attn_mask` specified - should call `_masked_softmax` according to NT nested structure (does this case need to be supported?)

Ideally, we'd register appropriate kernels to the `NestedTensorCPU` / `NestedTensorCUDA` / `NestedTensorAutograd` dispatch keys on the `softmax` / `_masked_softmax` ops as needed. Note that NT already supports standard `softmax` that works according to its nested structure.

If we absolutely need in-place softmax for perf, I'd expect that we implement this for NT underneath a `softmax_` op (not sure why this in-place variant doesn't exist already). Mutation semantics should not in general vary across device, but rather across op variants.

Things we should avoid:
* Introducing new NT-specific ops, even if they're private. It should be possible to register the NT keys for the actual ops (e.g. to `softmax` directly instead of creating a `_nested_tensor_softmax` op).
    * In practice, tracers bake these private calls into graphs and then we're subject to BC issues.
* Different mutation semantics for the same op across devices. Ops should have clearly-specified mutation semantics that match the naming conventions in PyTorch.

cc @albanD for additional thoughts on the design

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -12765,6 +12765,11 @@
     NestedTensorCPU: NestedTensor_to_padded_tensor_generic
     NestedTensorCUDA: NestedTensor_to_padded_tensor_cuda
 
+- func: _nested_tensor_softmax_with_shape(Tensor self, Tensor query) -> Tensor
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:12768]: My immediate reaction to this is that, NestedTensor_softmax_dropout is returning a view of the self, and therefore this is an incorrect schema. When this PR lands: https://github.com/pytorch/pytorch/pull/83371, this function would be composite implicit and instead of getting the buffer with `get_buffer` it would call `to_buffer`. I think the indexing loop will correctly register the backward formulas, but thats the only thing I would want to confirm with some grad_check tests.
